File: src/main/java/com/swelms/java/common/result/Result.java
package com.swelms.java.common.result;

import java.util.function.Function;
import java.util.function.Supplier;

public sealed interface Result<T> permits Result.Success, Result.Fail {
    T value();
    Throwable error();
    record Success<T>(T value) implements Result<T> {
        public Throwable error() { return null; }
    }

    record Fail<T>(Throwable error) implements Result<T> {
        public T value() { return null; }
    }

    static <T> Result<T> of(Supplier<T> block) {
        try {
            return new Success<>(block.get());
        } catch (Throwable e) {
            return new Fail<>(e);
        }
    }

    default Result<T> orElse(Function<Fail<T>, T> fallback) {
        return switch (this) {
            case Success<T> s -> s;
            case Fail<T> f -> of(() -> fallback.apply(f));
        };
    }


}

File: src/main/kotlin/com/swelms/common/result/Result.kt
package com.swelms.common.result

sealed interface Result<out T> {
   val value: T?
   operator fun component1(): T? = value
   class Success<T>(override val value: T) : Result<T>
   class Fail(val error: Throwable): Result<Nothing> {
      override val value get() = null
   }
}


fun <T> Result<T>.toKotlinResult(): kotlin.Result<T> =
   when (this) {
      is Result.Success -> kotlin.Result.success(value)
      is Result.Fail   -> kotlin.Result.failure(error)
   }

fun <T> kotlin.Result<T>.toSwelmsResult(): Result<T> =
   fold(
      onSuccess = { Result.Success(value = it) },
      onFailure = { Result.Fail(error = it) }
   )